File: sem3/map/Lab/2/src/model/ArithExp.java
package model;

public class ArithExp extends Exp {
    public static final int ADD = 1;
    public static final int SUBTRACT = 2;
    public static final int MULTIPLY = 3;
    public static final int DIVIDE = 4;
    private Exp e1, e2;
    private int op;
    public ArithExp(char op, Exp e1, Exp e2) {
        this.e1 = e1;
        this.e2 = e2;
        if(op == '+')
            this.op = ADD;
        else if(op == '-')
            this.op = SUBTRACT;
        else if(op == '*')
            this.op = MULTIPLY;
        else if(op == '/')
            this.op = DIVIDE;
    }

    @Override
    public int eval(MyIDictionary<String, Integer> symTable) {
        int rez = 0;
        int rez1 = this.e1.eval(symTable);
        int rez2 = this.e2.eval(symTable);
        if(op == ADD) {
            rez = rez1 + rez2;
        } else if(op == SUBTRACT) {
            rez = rez1 - rez2;
        } else if(op == MULTIPLY) {
            rez = rez1 * rez2;
        } else if(op == DIVIDE) {
//            if(rez2 == 0)
//                throw new DivideByZeroException("DivideByZeroException at: " + this.toString());
            rez = rez / rez2;
        }
        return rez;
    }

    @Override
    public String toString() {
        String ret = this.e1.toString();
        if(op == ADD)
            ret += " + ";
        else if(op == SUBTRACT)
            ret += " - ";
        else if(op == MULTIPLY)
            ret += " * ";
        else if(op == DIVIDE)
            ret += " / ";
        ret += this.e2.toString();
        return ret;
    }
}

File: sem3/sem 3/~craciunf/c-sharp-toy-language-interpreter/models/ArithExp.cs
/* package models; */
using System;
using System.Collections;
using System.Collections.Generic;

namespace ToyLanguageInterpreter {
    public class ArithExp : Exp {
        private Exp e1, e2;
        private char op;

        public ArithExp(Exp e1, Exp e2, char op) {
            this.e1 = e1;
            this.e2 = e2;
            this.op = op;
        }

        /* @Override */
        public override int eval(Dictionary <String, int> symTable) {
            int partialRes1 = this.e1.eval(symTable);
            int partialRes2 = this.e2.eval(symTable);

            switch(this.op) {
                case '+': return partialRes1 + partialRes2;
                case '-': return partialRes1 - partialRes2;
                case '*': return partialRes1 * partialRes2;
                case '/': return partialRes1 / partialRes2;
            }

            return 0;
        }

        /* @Override */
        public override String toString() {
            return this.e1.toString() + " " + this.op + " " + this.e2.toString();
        }
    }
}
